### PR TITLE
Typed Layout for easy conversion from Layout to actual value

### LINKED ISSUE #1811: Layoutable non-string values
Current approach:

render, parse.

e.g.

`Convert.ToInt64(this.MaxKilobytes.Render(logEvent), CultureInfo.InvariantCulture);`

This isn't optimal. 

Proposal, replace Layout with `Layout<int>` ~`IntLayout`,~ and let `render` do a parse.

related #1234 

update: We have now also an implicit conversion from `string` to `layout`. For the `int` case it should be also from `int` to `layout` (fixed values), hence we need `IntLayout` or something like that.

### LINKED ISSUE #3994: Extend typed layout so it recognizes string parses
Follow up on #3656 

Extend typed layout so it recognizes string parses

Currently this doesn't work:

```c#
[Fact]
public void ComplexTypeTestWithStringConversion()
{
    // Arrange
    var value = "utf8";
    var layout = CreateLayoutRenderedFromProperty<Encoding>();
    var logEventInfo = CreateLogEventInfoWithValue(value);

    // Act
    var result = layout.RenderToValueOrDefault(logEventInfo);

    // Assert
    Assert.Equal("utf8", result.EncodingName);
}
 ```

error:

```        
System.InvalidCastException
Invalid cast from 'System.String' to 'System.Text.Encoding'.
   at System.Convert.DefaultToType(IConvertible value, Type targetType, IFormatProvider provider)
   at NLog.Common.ValueConverter.Convert(Object propertyValue, Type propertyType, String format, IFormatProvider formatProvider) in D:\nlog\NLog\src\NLog\Common\ValueConverter.cs:line 91
   at NLog.Layouts.Layout`1.TryConvertTo(Object raw, T& value) in D:\nlog\NLog\src\NLog\Layouts\Typed\Layout.cs:line 129
   at NLog.Layouts.Layout`1.TryConvertRawToValue(Object raw, T& value) in D:\nlog\NLog\src\NLog\Layouts\Typed\Layout.cs:line 266
   at NLog.Layouts.Layout`1.TryGetRawValueIntern(LogEventInfo logEvent, T& rawValue) in D:\nlog\NLog\src\NLog\Layouts\Typed\Layout.cs:line 237
   at NLog.Layouts.Layout`1.RenderToValueInternal(LogEventInfo logEvent, StringBuilder reusableBuilder) in D:\nlog\NLog\src\NLog\Layouts\Typed\Layout.cs:line 206
   at NLog.Layouts.Layout`1.NLog.Layouts.IRenderable<T>.RenderToValue(LogEventInfo logEvent) in D:\nlog\NLog\src\NLog\Layouts\Typed\Layout.cs:line 187
   at NLog.Layouts.RenderableExtensions.RenderToValueOrDefault[T](Layout`1 l, LogEventInfo logEvent, T defaultValue) in D:\nlog\NLog\src\NLog\Layouts\Typed\RenderableExtensions.cs:line 60
   at NLog.UnitTests.Layouts.Typed.LayoutTests.ComplexTypeTestWithStringConversion() in D:\nlog\NLog\tests\NLog.UnitTests\Layouts\Typed\LayoutTests.cs:line 170
```

while string to int does work.

The logic for converting string to Encoding (and other stuff) is inside `PropertyHelper`

### PR DESCRIPTION
**Pull request by snakefoot**
Completed my own version of #3656. That resolves #1811 and resolves #3994

With full optimization of layout-precalculation and caching of previous value to reduce boxing.

Created my own version because I wasn't sure #3656 would be completed. Maybe steal the precalculate-logic from this PR ?

Notice that I have decided to always use the `Layout<T>` type-converter-abilities, instead of depending on `PropertyHelper`. This makes the behavior more predictable for users creating their own `Layout<T>`.

### CODE REVIEW

**Comment by snakefoot** [comment]
Used the `Layout<T>` for creating a helper-class for implementing `ValueType`-property just like for DatabaseTarget-Parameters.

The goal is to add ValueType to JsonLayout-Attribute, so one can guide the wanted formatting. Instead of using `Encode="false" as a trick to handle integer-output.

File: src/NLog/Targets/MailTarget.cs [hunk 1]
```diff
@@ -636,21 +636,7 @@ private MailMessage CreateMailMessage(LogEventInfo lastEvent, string body)
...
-                else if (ConversionHelpers.TryParseEnum(renderedPriority, out MailPriority mailPriority))
-                {
-                    msg.Priority = mailPriority;
-                }
-                else
-                {
-                    msg.Priority = MailPriority.Normal;
-                    InternalLogger.Warn("{0}: Could not convert '{1}' to MailPriority, valid values are Low, Normal and High. Using normal priority as fallback.", this, renderedPriority);
-                }
+                msg.Priority = Priority.RenderValue(lastEvent, MailPriority.Normal);
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/MailTarget.cs:639]: Yes if would fail with NullReferenceException, if not for the check just above `if (Priority != null)`

**Comment by snakefoot** [comment]
@304NotModified Would like to have #3656 or this PR merged, so I can continue with extracting NLog.DatabaseTarget from NLog-core. Any favorite direction?

**Comment by 304NotModified** [comment]
> @304NotModified Would like to have #3656 or this PR merged, so I can continue with extracting NLog.DatabaseTarget from NLog-core. Any favorite direction?

I really like to merge one of them, but haven't checked this one in depth

File: src/NLog/Targets/MailTarget.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/Targets/MailTarget.cs:639]: And what about the `RenderToValueOrDefault` and hiding renderValue? I'm really afraid we get a lot of nullref exceptions in many (3th party) libraries

**Comment by 304NotModified** [comment]
Is the conclusion for this PR that `Layout<T>` has limited support for `T`? Should we throw an exception if `T` isn't supported?

File: src/NLog/Targets/MailTarget.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/MailTarget.cs:639]: Yes excellent idea with the extension-method. Just need to be in the NLog-namespace.

**Comment by snakefoot** [comment]
> Is the conclusion for this PR that `Layout<T>` has limited support for `T`? Should we throw an exception if `T` isn't supported?

Yes I guess so, until a champion arrives with lots of time and great ideas :)

Maybe start with `T` of type `ICollection` ?

**Comment by 304NotModified** [comment]
> Maybe start with throwing when `T` is of type `ICollection` ?

Shouldn't be a white list?

File: src/NLog/Targets/Target.cs [hunk 2]
```diff
@@ -710,6 +710,36 @@ protected string RenderLogEvent(Layout layout, LogEventInfo logEvent)
...
 
+        /// <summary>
+        /// Renders the logevent into a result-value by using the provided layout
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="layout">The layout.</param>
+        /// <param name="logEvent">The logevent info.</param>
+        /// <param name="defaultValue">Fallback value when no value available</param>
+        /// <returns>Result value when available, else fallback to defaultValue</returns>
+        protected T RenderLogEvent<T>(Layout<T> layout, LogEventInfo logEvent, T defaultValue = default(T))
```
Comment (304NotModified) [inline src/NLog/Targets/Target.cs:721]:
```suggestion
        protected T RenderLogEvent<T>([CanBeNull] Layout<T> layout, [CanBeNull] LogEventInfo logEvent, T defaultValue = default(T))
```

File: src/NLog/Targets/WebServiceTarget.cs [hunk 3]
```diff
@@ -255,7 +243,7 @@ protected override void DoInvoke(object[] parameters)
         /// <param name="continuation">The continuation.</param>
         protected override void DoInvoke(object[] parameters, AsyncContinuation continuation)
         {
-            var url = (Protocol != WebServiceProtocol.HttpGet ? _fixedUri : null) ?? BuildWebServiceUrl(LogEventInfo.CreateNullEvent(), parameters);
+            var url = Protocol != WebServiceProtocol.HttpGet ? Url.StaticValue : BuildWebServiceUrl(LogEventInfo.CreateNullEvent(), parameters);
```
Comment (304NotModified) [inline src/NLog/Targets/WebServiceTarget.cs:246]: I don't get this, why do we use the StaticValue? Also Url could here be `null`?

File: tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using NLog.Layouts;
+using Xunit;
+
+namespace NLog.UnitTests.Layouts
+{
+    public class LayoutTypedTests : NLogTestBase
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs:40]: are these new tests, or taken from #3656?

Comment (snakefoot) [reply tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs:40]: These are new. Implemented to verify the internal caching.

File: src/NLog/Targets/WebServiceTarget.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Targets/WebServiceTarget.cs:246]: `Url` is a required property, so the target cannot initialize without it being assigned.

**Comment by snakefoot** [comment]
> Shouldn't be a white list?

Anything that is immutable and can be understood by `IPropertyTypeConverter` is supported. But I don't mind a white-list, but I have a feeling that people suddenly want to modify the white-list.

File: src/NLog/Targets/Target.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/Targets/Target.cs]: Fixed

File: src/NLog/Targets/MailTarget.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/MailTarget.cs]: Fixed

File: tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs (hunk 4 above)
Comment (snakefoot) [reply tests/NLog.UnitTests/Layouts/LayoutTypedTests.cs:40]: But I have now included the tests from #3656

**Comment by snakefoot** [comment]
Rebased to dev-branch because change to VS2019-pipeline caused PullRequest on VS2017-pipeline to fail.

**Comment by snakefoot** [comment]
Rebased to dev-branch because conflicting change from #4333

**Comment by snakefoot** [comment]
Added type-validation as part of the `Layout<T>`-constructor.

**Pull request merged**